File: chrome/android/java/src/org/chromium/chrome/browser/metrics/UkmRecorder.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.metrics;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.content_public.browser.WebContents;

/**
 * An interface and classes to record User Keyed Metrics.
 */
public abstract class UkmRecorder {
    /**
     * Records the occurrence of a (boolean) UKM event with name |eventName|.
     * A UKM entry with |eventName| must be present in ukm.xml with a metric named `HasOccured`.
     * For example,
     * <event name="SomeFeature.SomeComponent">
     * <owner>owner@chromium.org</owner>
     * <summary>
     * User triggered a specific feature.
     * </summary>
     * <metric name="HasOccurred" enum="Boolean">
     * <summary>
     * A boolean signaling that the event has occurred (typically only records
     *  true values).
     * </summary>
     * </metric>
     * </event>
     */
    abstract void recordEvent(WebContents webContents, String eventName);

    /**
     * The actual recorder.
     */
    @JNINamespace("metrics")
    public static class Bridge extends UkmRecorder {
        @Override
        public void recordEvent(WebContents webContents, String eventName) {
            UkmRecorderJni.get().recordEvent(webContents, eventName);
        }
    }

    @NativeMethods
    interface Natives {
        void recordEvent(WebContents webContents, String eventName);
    }
}

File: chrome/browser/android/metrics/ukm_recorder.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "services/metrics/public/cpp/ukm_recorder.h"
#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "chrome/android/chrome_jni_headers/UkmRecorder_jni.h"
#include "components/ukm/content/source_url_recorder.h"
#include "content/public/browser/web_contents.h"
#include "services/metrics/public/cpp/ukm_builders.h"
#include "services/metrics/public/cpp/ukm_entry_builder.h"

namespace metrics {

// Called by Java org.chromium.chrome.browser.metrics.UkmRecorder.
static void JNI_UkmRecorder_RecordEvent(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& j_web_contents,
    const base::android::JavaParamRef<jstring>& j_event_name) {
  content::WebContents* web_contents =
      content::WebContents::FromJavaWebContents(j_web_contents);
  const ukm::SourceId source_id =
      ukm::GetSourceIdForWebContentsDocument(web_contents);
  const std::string event_name(ConvertJavaStringToUTF8(env, j_event_name));
  ukm::UkmEntryBuilder builder(source_id, event_name);
  builder.SetMetric("HasOccurred", true);
  builder.Record(ukm::UkmRecorder::Get());
}

}  // namespace metrics
